staging/kos/pkg/controllers/subnet/validator.go

```diff
@@ -0,0 +1,524 @@
+// to the Validator.
+func (v *Validator) OnSubnetUpdate(oldObj, newObj interface{}) {
+	oldS, newS := oldObj.(*netv1a1.Subnet), newObj.(*netv1a1.Subnet)
+	glog.V(5).Infof("Notified of update from %#+v to %#+v.", oldS, newS)
+	sRef := k8stypes.NamespacedName{
+		Namespace: newS.Namespace,
+		Name:      newS.Name,
+	}
+	// Process a subnet only if the fields that affect validation have changed.
+	if oldS.Spec.IPv4 != newS.Spec.IPv4 || oldS.Spec.VNI != newS.Spec.VNI {
+		v.queue.Add(sRef)
+	}
+}
+
+// OnSubnetDelete is invoked when a subnet is removed from the Validator's
+// informer's cache. It enqueues a reference to the subnet.
+func (v *Validator) OnSubnetDelete(obj interface{}) {
+	s := kosctlrutils.Peel(obj).(*netv1a1.Subnet)
+	glog.V(5).Infof("Notified of deletion of %#+v.", s)
+	sRef := k8stypes.NamespacedName{
+		Namespace: s.Namespace,
+		Name:      s.Name,
+	}
+	v.queue.Add(sRef)
+}
+
+func (v *Validator) processQueue() {
+	for {
+		subnet, stop := v.queue.Get()
+		if stop {
+			return
+		}
+		v.processQueueItem(subnet.(k8stypes.NamespacedName))
+	}
+}
+
+func (v *Validator) processQueueItem(subnet k8stypes.NamespacedName) {
+	defer v.queue.Done(subnet)
+	err := v.processSubnet(subnet.Namespace, subnet.Name)
+	requeues := v.queue.NumRequeues(subnet)
+	if err == nil {
+		glog.V(4).Infof("Finished %s with %d requeues.", subnet, requeues)
+		v.queue.Forget(subnet)
+		return
+	}
+	glog.Warningf("Failed processing %s, requeuing (%d earlier requeues): %s.", subnet, requeues, err.Error())
+	v.queue.AddRateLimited(subnet)
+}
+
+func (v *Validator) processSubnet(namespace, name string) error {
+	s, err := v.subnetLister.Subnets(namespace).Get(name)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		glog.Errorf("subnet lister failed to lookup %s/%s: %s", namespace, name, err.Error())
+		// This should never happen. No point in retrying.
+		return nil
+	}
+	if k8serrors.IsNotFound(err) {
+		v.processDeletedSubnet(namespace, name)
+		return nil
+	}
+	return v.processExistingSubnet(s)
+}
+
+func (v *Validator) processDeletedSubnet(namespace, name string) {
+	rivals := v.clearConflictsCache(k8stypes.NamespacedName{
+		Namespace: namespace,
+		Name:      name,
+	})
+	// re-validate old rivals: they might no longer have conflicts as this
+	// subnet has been deleted.
+	for _, r := range rivals {
+		v.queue.Add(r)
+	}
+}
+
+func (v *Validator) processExistingSubnet(s *netv1a1.Subnet) error {
+	sd, err := parseSubnet(s)
+	if err != nil {
+		return err
+	}
+	nsn := k8stypes.NamespacedName{
+		Namespace: s.Namespace,
+		Name:      s.Name,
+	}
+	// If we're here s has either been created or updated in a way that affects
+	// validation. We need to update its conflicts cache accordingly.
+	oldRivals := v.updateConflictsCache(nsn, sd)
+	// Reconsider old rivals because s has changed and they might no longer be
+	// in conflict.
+	for _, r := range oldRivals {
+		v.queue.Add(r)
+	}
+
+	if s.Status.ValidationOutcome == usable {
```

Comment:
The pattern of keeping implementation state in Status is inferior.  In this case it can lead to changes in `s.Status.ValidationOutcome` that we probably do not want.  Consider what happens if a Subnet is processed again before the informer's cache is notified about the update resulting from previous processing.

Reply:
I will discuss this comment in the thread about  "three conflicts lists vs a single list", as many things that are relevant to this comment as well have been said in that thread.